trinity/sync/full/chain.py

```diff
@@ -448,7 +450,17 @@ def register_peer(self, peer: BasePeer) -> None:
 
             try:
                 self._block_persist_tracker.register_tasks(headers)
-            except ValidationError:
+            except DuplicateTasks as exc:
+                # Likely scenario: switched which peer downloads headers, and the new peer isn't
+                # aware of some of the in-progress headers
+                self.logger.debug("Duplicate headers during fast sync %r, skipping", exc.duplicates)
+                # Tried a bunch of things that failed to make mypy happy:
+                # - cast(DuplicateTasks[BlockHeader], exc) -- refuses to cast
+                # - except DuplicateTasks[BlockHeader]: -- doesn't catch the exception
+                # - OrderedTaskPreparation.DuplicateTasks to inherit TTask -- mypy refuses
```

Comment:
Can you add a line up above the `try`:

```python
exc: DuplicateTask[BlockHeader]
try:
    ...
except DuplicateTasks as exc:
   ...
```

I think that's how you tell `mypy` *ahead of time* what type something is.

Reply:
Pre-defining the exception type still complained... going with the `duplicates` cast.

Reply:
I rabbit hole'd this a while ago.  There's a solution that works for simple types but not this one but which has a runtime cost.

```python
if isinstance(v, SomeType):
    do_SomeType_specific_thing(v)
```

`mypy` correctly understands that within the `if` context `v` **must** be `SomeType`.  However, there's no way to do the equivalent of `if isinstance(v, Tuple[SomeType, ...])` that I could find.  :cry: